R/update-input.R

```diff
@@ -644,7 +644,7 @@ updateSelectInput <- function(session = getDefaultReactiveDomain(), inputId, lab
   choices <- if (!is.null(choices)) choicesWithNames(choices)
   if (!is.null(selected)) selected <- as.character(selected)
   options <- if (!is.null(choices)) selectOptions(choices, selected, inputId, FALSE)
-  message <- dropNulls(list(label = label, options = options, value = selected))
+  message <- dropNulls(list(label = label %convert% as.character, options = options, value = selected))
```

Comment:
This should probably be using `processDeps()` (so that `htmlDependency()`s don't get dropped), and `Shiny.renderDependency()` to render them client-side.

Also, if we think `label` has the potential to contain bindings, we should be using `Shiny.renderContent()` instead of `.html()`

Reply:
Ah yes, that makes sense, thanks @cpsievert !

I've pushed something that does just that but it's a bit more changes than anticipated since `renderContent` is async everything upstream and downstream has to be slightly changed (e.g.: `receiveMessage` on the input binding).

This now works.

```r
library(shiny)

ui <- fluidPage(
  textInput("text", span("Text", class = "text-danger"), "Hello, world!"),
  actionLink("update", "update label")
)

server <- function(input, output, session) {
  observeEvent(input$update, {
    updateTextInput(session, "text", label = span("Updated", icon("eye"), class = "text-success"))
  }) 
}

shinyApp(ui, server)
```